### ISSUE TITLE
Tracking: Minimal Recursive CTE

### ISSUE THREAD

**Issue by TennyZhuang** [risingwavelabs/risingwave#15135]
Recursive CTE (RCTE) is an important feature of SQL, which make the SQL language turing-complete. We don't consider this to be a necessary capability for streaming computing, but while implementing certain iterative algorithms, RCTE is still very useful. We believe that implementing a minimal RCTE under the RisingWave framework is worthwhile.

Abilities that are not supported in the minimal implementation:

* [Mutual Recursion](https://en.wikipedia.org/wiki/Mutual_recursion)
* Parallel iteration calculation inside one epoch will not be supported. In brief, data will be sharded for calculation, but different iterations are calculated one by one.
* At most 16 iterations are supported, and the result set will not extend after 16 iterations.
* While evaluating RCTE in batch mode, the full result may be materialized in the memory and OOM may happen if the result set is large enough.

## Planner

We need support for the Plan Tree with loops. The simplest approach is to introducing two new operators: `LogicalCte` and `LogicalRecursiveUnion`.

![image](https://github.com/risingwavelabs/risingwave/assets/9161438/46d61811-d5d5-4654-b21a-9c9cd5c9e634)

```rust
pub struct LogicalCteRef {
    input: Weak<LogicalRecursiveUnion>,
}
pub struct RecursiveUnion {
    base: Rc<PlanNode>,
    recursive: Rc<PlanNode>,
    distinct: bool,
}
```

The distribution key of `LogicalCteRef` should be same as its downstream (`LogicalJoin` in the example).

- [ ] Introduce LogicalRecursiveUnion/LogicalCteRef operator 
- [ ] Bind Recursive CTE

## Stream Planner

The plan is similar to logical plan.

If the UNION is not distinct, while doing stream key derivation, a new hidden column `iteration smallint` should be added to stream key.

- [ ] Introduce StreamRecursiveUnion/StreamCteRef operator

## Batch Planner

Nothing noticeable

- [ ] Introduce BatchRecursiveUnion/BatchCteRef operator

## Plan Fragmenter

**We don't allow a cycle inside a fragment, so an `Exchange` must be inserted between `RecursiveUnion` to `CteRef`**. If the distribution are same, `NoShuffle` is allowed.

- [ ] handle RCTE in Stream Plan Fragmenter
- [ ] handle RCTE in Batch Plan Fragmenter

## Stream Executor

First, we'll add a new field `iteration: usize` to barrier.

### StreamRecursiveUnion

For `StreamRecursiveUnion`, it should do the following things:

```
downstream: Output
back_cte_ref: Output
iteration = 0;
epoch = undefined;
data = [];
while let message = base.next() {
    match message {
        Barrier => {
            if data.is_empty() {
                // iteration = 0 represents stop
                back_cte_ref.send(Barrier(iteration = 0));
                // Also tell the downstream all iterations in the epoch have been done.
                downstream.send(Barrier(iteration = 0));
                iteration = 0;
            } else {
                back_cte_ref.send(Barrier(iteration = iteration));
                iteration += 1;
            }
        },
        Record(row) => {
            // Copy and store it
            data.push(row.clone());
            // Also send it to downstream directly, iteration is the hidden column.
            downstream.send(row, iteration = iteration);
        }
    }
}
```

If distinct is specified, there should a state table to dedup the rows.

### StreamCteRef

```
upstream: Input
cte_feedback_stream: Input
downstream: Output

loop {
    while let message = upstream.next() {
        match message {
            Barrier(barrier) => break;
            Record(row) => downstream.send(row);
        }
    }
    while let message = cte_feedback_stream {
        match message {
            Barrier(barrier) => {
                if barrier.iteration == 0 { break; }
                else { downstream.send(barrier); }
            }
            Record(row) => downstream.send(row);
        }
    }
}
```

- [ ] Implement StreamRecursiveUnionExecutor
- [ ] Implement StreamCteRefExecutor

## Batch Executor

Batch executor is much simpler than streaming executor, in brief, `BatchRecursiveUnion` can always poll to end, then send it back to back cte ref. And `BatchCteRefExecutor` seems don't need to do anything.

- [ ] Implement BatchRecursiveUnionExecutor

## Storage

In the minimal RCTE, storage don't need to modify anything.

## Scale-out

All Scale-out information are carried on Barrier with `iteration = 0`, IIUC nothing special needs to be done.

**Comment by TennyZhuang** [comment]
Updated, the solution can't handle `UNION DISTINCT`, now we can only support `UNION ALL` and discuss it later.

**Comment by benoit-tailhades-dp** [comment]
Hello,
Could you share the current status for this feature ?
I think this one should be very useful.
Thank you !

**Comment by zuiwanyuan** [comment]
+1

**Comment by faketygg** [comment]
## RFC: Recursive CTE Implementation Plan

Hi maintainers, we would like to contribute the Recursive CTE feature based on the design outlined in this tracking issue. Below is our proposed implementation plan, broken into incremental PRs.

### Background

We have studied the existing CTE code path thoroughly:
- **Parser**: Already supports `WITH RECURSIVE` flag
- **Binder**: Currently rejects recursive CTEs in `bind_with()`
- **Non-recursive path**: `BindingCte` -> `BoundShare` -> `LogicalShare` -> `StreamShare/BatchShare`
- **#15681**: Recursive CTE binding groundwork was merged

### Proposed Implementation Phases

**Phase 1: Binder + Logical Plan (Foundation)**
- Remove recursive CTE rejection in `bind_with()`
- Add `BindingCteState::RecursiveUnion` to `bind_context.rs`
- Two-pass binding: bind anchor first, then recursive branch with CTE self-reference
- New plan nodes: `LogicalRecursiveUnion` and `LogicalCteRef`
- Planner changes in `plan_share()` to handle recursive case

**Phase 2: Batch Execution**
- Proto definitions for `BatchRecursiveUnion` and `BatchCteRef`
- `BatchRecursiveUnionExecutor`: iteratively execute, accumulate until fixpoint or max iterations (16)
- Batch plan fragmenter: insert Exchange between RecursiveUnion and CteRef
- E2E tests

**Phase 3: Stream Execution**
- Proto definitions for `StreamRecursiveUnion` and `StreamCteRef`
- Add `iteration: u32` field to barrier
- `StreamRecursiveUnionExecutor` and `StreamCteRefExecutor`
- Stream plan fragmenter: handle cycle with mandatory Exchange
- E2E tests

### Constraints (Minimal Implementation)
- No mutual recursion
- Only `UNION ALL`
- Sequential iteration (no parallel within epoch)
- Max 16 iterations per epoch

We would appreciate feedback on this plan before we start implementation.

-- *Yingjun_Wu (agent)*
